src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -23,160 +23,47 @@ object LoadMatrix {
-        newoff += 1
-      }
-      if (newoff - off == missingValue.length && (string.length == newoff || string(newoff) == sep)) {
-        rvb.setMissing()
-        newoff
-      } else {
-        fatal(s"Error parsing matrix. Invalid Int64 at column: $colNum, row: $rowID in file: $file")
-      }
-    } else if (string.length == newoff || string(newoff) == sep) {
-      rvb.addLong(if (isNegative) -v else v)
-      newoff
-    } else {
-      fatal(s"Error parsing matrix. Invalid Int64 at column: $colNum, row: $rowID in file: $file")
-    }
-  }
-
-  def setFloat(string: String, off: Int, rvb: RegionValueBuilder, sep: Char, missingValue: String, file: String, rowID: String, colNum: Int): Int = {
-    var newoff = string.indexOf(sep, off)
-    if (newoff == -1) {
-      newoff = string.length
-    }
-    val v = string.substring(off, newoff)
-    if (v == missingValue) {
-      rvb.setMissing()
-    } else {
-      try {
-        rvb.addFloat(v.toFloat)
-      } catch {
-        case _: NumberFormatException => fatal(s"Error parsing matrix: $v is not a Float32. column: $colNum, row: $rowID in file: $file")
-      }
-    }
-    newoff
-  }
-
-  def setDouble(string: String, off: Int, rvb: RegionValueBuilder, sep: Char, missingValue: String, file: String, rowID: String, colNum: Int): Int = {
-    var newoff = string.indexOf(sep, off)
-    if (newoff == -1) {
-      newoff = string.length
-    }
-    val v = string.substring(off, newoff)
-    if (v == missingValue) {
-      rvb.setMissing()
-    } else {
-      try {
-        rvb.addDouble(v.toDouble)
-      } catch {
-        case _: NumberFormatException => fatal(s"Error parsing matrix: $v is not a Float64! column: $colNum, row: $rowID in file: $file")
-      }
-    }
-    newoff
-  }
-
   def apply(hc: HailContext,
     files: Array[String],
+    nAnnotations: Int = 1,
+    annotationHeaders: Option[Seq[String]],
+    annotationTypes: Seq[Type],
+    keyExpr: String,
     nPartitions: Option[Int] = None,
     dropSamples: Boolean = false,
     cellType: Type = TInt64(),
-    sep: String = "\t",
-    missingValue: String = "NA",
-    hasRowIDName: Boolean = false): VariantSampleMatrix = {
-
-    val cellParser: (String, Int, RegionValueBuilder, Char, String, String, String, Int) => Int = cellType match {
-      case _: TInt32 => setInt
-      case _: TInt64 => setLong
-      case _: TFloat32 => setFloat
-      case _: TFloat64 => setDouble
-      case _: TString => setString
-      case _ =>
-      fatal(
-        s"""expected cell type Int32, Int64, Float32, Float64, or String but got:
-           |    ${ cellType.toPrettyString() }
-         """.stripMargin)
+    seperator: String = "\t",
+    missingValue: String = "NA"): VariantSampleMatrix = {
+
+    assert(annotationTypes.length == nAnnotations)
+    for (t <- annotationTypes) {
+      assert(t.isOfType(TString()) || t.isOfType(TInt32()) || t.isOfType(TInt64()) || t.isOfType(TFloat32()) || t.isOfType(TFloat64()))
```

Comment:
Needs error message.  Might be easier to test in Python (in which case you can leave assert).

If you leave assert, all the code related to the assert should be in the assert, so break it out into a separate function or do something like `assert(annotationTypes.forall(t => t.isOfType(TString()) || ...)`.